src/io/parsers/GROParser.js

```diff
@@ -0,0 +1,230 @@
+   */
+  _parseAtom(line) {
+    this._residueNumber = line.readInt(1, 5);
+    this._residueName = line.readString(6, 10).trim();
+    this._atomName = line.readString(11, 15).trim();
+    this._atomNumber = line.readInt(16, 20);
+    const positionX = line.readFloat(21, 28) * 10;
+    const positionY = line.readFloat(29, 36) * 10;
+    const positionZ = line.readFloat(37, 45) * 10;
+    if (Number.isNaN(positionX) || Number.isNaN(positionY) || Number.isNaN(positionZ)) {
+      this._complex.error = {
+        message: `Atom position is invalid in "${line.readLine()}"`,
+      };
+      return;
+    }
+    /* const velocityX = line.readFloat(46, 53);
+    const velocityY = line.readFloat(54, 61);
+    const velocityZ = line.readFloat(62, 69); */
+    /* Adding residue and atom to complex structure */
+    const type = Element.getByName(this._atomName[0]); /* MAGIC 0. REASONS: This name is something like "CA", where
+     C - is an element an A is something else. But what about Calcium? */
+    if (type.fullName === 'Unknown') {
+      this._complex.error = {
+        message: `${this._atomName[0]} hasn't been recognised as an atom name.`,
+      };
+      return;
+    }
+    const role = Element.Role[this._atomName];
+    /* Firstly, create a dummy chain */
+    let chain = this._chain;
+    if (!chain) {
+      this._chain = chain = this._complex.addChain('A');
+    }
+    /* Secondly, add residue to that chain */
+    let residue = this._residue;
+    if (!residue || residue.getSequence() !== this._residueNumber) {
+      this._residue = residue = chain.addResidue(this._residueName, this._residueNumber, 'A');
+    }
+    /* Lastly, add atom to that residue */
+    this._atomPosition = new THREE.Vector3(positionX, positionY, positionZ);
+    /* Adding default constants to correct atom addition process */
+    const het = true;
+    const altLoc = ' ';
+    const occupancy = 1;
+    const tempFactor = 1;
+    const charge = 0;
+    residue.addAtom(this._atomName, type, this._atomPosition, role, het, this._atomNumber, altLoc, occupancy, tempFactor, charge);
+  }
+
+  /**
+   * Needed procedure for molecules finalization. In '.gro' file format there is only 1 chain and 1 molecule.
+   */
+  _finalizeMolecules() {
+    // get chain from complex
+    const chainDict = {};
+    const chain = this._complex._chains[0];
+    const chainName = chain._name;
+    chainDict[chainName] = chain;
+
+    // aggregate residues from chain
+    const m = this._molecules[0];
+    let residues = [];
+    residues = residues.concat(chain._residues.slice());
+
+    const molecule = new Molecule(this._complex, m._name, 1);
+    molecule._residues = residues;
+    this._complex._molecules[0] = molecule;
+  }
+
+  /**
+   * Some finalizing procedures.
+   * @returns {Complex} Complex structure for visualizing.
+   */
+  _finalize() {
+    this._molecule = { _index: '', _chains: [] };
+    this._molecule._index = 1;
```

Comment:
What's the point with such a strange approach to molecules in this and the following lines? Is it a kind of copy-paste from a more general case? Should we refactor/optimize this code?

Reply:
We need to create molecules from scratch, and this is some finalizing procedures. Originally it was some sort of copy-paste from PDB Parser, but with adaptation to specific gro file format features.

Reply:
I mean, look at the lines 166 and 167, do they make sense for you? Why are they written separately in such a strange way? Should we merge them? Will you do it? :)

Another concern is the call to _finalizeMolecules() which pretends to be written in a generic way but allows only for one chain and one molecule. What's the point for writing such ridiculous code? lines 146-149 could be merged into at most two lines; lines 153-153 (+157) are overkill... The entire function can be rewritten in a clear and concise way.

Another step would be to merge _finalizeMolecules() into _finalize() and re-optimize again making it more clear. If we know that there's only one molecule and one chain then we can do it. Should do it. Must :)

Reply:
Please take a look on new edition. I guess we can left finalizing in that state, because all we need is 2-3 simple lines (creating molecule, adding chains and residues and that's all). So i guess there is no need for descrete function for this (only in GROParser case).